Optimize aten.cat calls of a repeated element

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132081

This was a particular problem for a model I saw which would have a large number of repeats, making compilation slow.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Optimize aten.cat calls of a repeated element

[ghstack-poisoned]

eellison committed:
Update on "Optimize aten.cat calls of a repeated element"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Optimize aten.cat calls of a repeated element"


This was a particular problem for a model I saw which would have a large number of repeats, making compilation slow.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented on torch/_inductor/decomposition.py:
```diff
@@ -350,6 +350,17 @@ def non_empty_tensor(x: torch.Tensor) -> bool:
         return aten.cat.default(filtered_tensors, dim)
+
+    # optimization, avoid concat for single, repeated input
+    if len(filtered_tensors) >= 1 and all(
```
nit: `len(filtered_tensors) > 1` , since `== 1` case has been handled above.

eellison committed:
Update on "Optimize aten.cat calls of a repeated element"


This was a particular problem for a model I saw which would have a large number of repeats, making compilation slow.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "Optimize aten.cat calls of a repeated element"


This was a particular problem for a model I saw which would have a large number of repeats, making compilation slow.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
